File: packages/geometry-utils/src/polygon.ts
import Point from './point';
import { almostEqual, cycleIndex } from './shared-helpers';
import { IPoint, IPolygon } from './types';

export default class Polygon {
    private innerChildren: Polygon[];

    private data: Float64Array;

    private offset: number;

    private pointCount: number;

    private closed: boolean;

    private closedDirty: boolean;

    private tmpPoint: Point;

    private rectangle: boolean;

    private rectData: Float64Array;

    private constructor() {
        this.tmpPoint = Point.zero();
        this.innerChildren = [];
        this.rectData = new Float64Array(4);
        this.closed = false;
        this.pointCount = 0;
        this.offset = 0;
        this.rectangle = false;
    }

    public reset(points: IPoint[]): void {
        this.pointCount = points.length;
        this.offset = 0;
        this.data = new Float64Array(this.getPointOffset(this.pointCount + 2));

        let i: number = 0;

        for (i = 0; i < this.pointCount; ++i) {
            this.tmpPoint.bind(this.data, this.getPointOffset(i)).update(points[i]);
        }

        this.calculateBounds();
    }

    public rotate(angle: number): void {
        const pointCount: number = this.pointCount;
        const radianAngle: number = (angle * Math.PI) / 180;
        let i: number = 0;

        for (i = 0; i < pointCount; ++i) {
            this.at(i).rotate(radianAngle);
        }

        this.calculateBounds();
    }

    public at(index: number): Point | null {
        if (index >= this.length) {
            return null;
        }

        const pointIndex: number = cycleIndex(index, this.pointCount, 0);

        return this.tmpPoint.bind(this.data, this.getPointOffset(pointIndex));
    }

    public pointIn(point: IPoint, offset: Point = null): boolean {
        if (this.isBroken) {
            return null;
        }

        const innerPoint: Point = Point.from(point);
        const currPoint: Point = Point.zero();
        const prevPoint: Point = Point.zero();
        const pointCount: number = this.pointCount;
        let inside: boolean = false;
        let i: number = 0;

        for (i = 0; i < pointCount; ++i) {
            currPoint.update(this.at(i));
            prevPoint.update(this.at(cycleIndex(i, pointCount, -1)));

            if (offset !== null) {
                currPoint.add(offset);
                prevPoint.add(offset);
            }

            //  no result                            exactly on the segment
            if (currPoint.almostEqual(innerPoint) || innerPoint.onSegment(currPoint, prevPoint)) {
                return null;
            }

            if (currPoint.almostEqual(prevPoint)) {
                // ignore very small lines
                continue;
            }

            if (
                currPoint.y > innerPoint.y !== prevPoint.y > innerPoint.y &&
                innerPoint.x < innerPoint.interpolateX(prevPoint, currPoint)
            ) {
                inside = !inside;
            }
        }

        return inside;
    }
    public close(): void {
        if (this.isClosed) {
            return;
        }

        this.closedDirty = true;
    }

    public reverse(): void {
        const halfPointCount: number = this.pointCount >> 1;
        const lastIndex: number = this.pointCount - 1;
        let i: number = 0;
        let j: number = 0;
        let i2: number = 0;
        let j2: number = 0;
        let i2Plus1: number = 0;
        let j2Plus1: number = 0;

        for (i = 0; i < halfPointCount; ++i) {
            j = lastIndex - i;
            i2 = this.offset + (i << 1);
            j2 = this.offset + (j << 1);
            i2Plus1 = this.offset + i2 + 1;
            j2Plus1 = this.offset + j2 + 1;

            this.data[i2] = this.data[i2] + this.data[j2];
            this.data[j2] = this.data[i2] - this.data[j2];
            this.data[i2] = this.data[i2] - this.data[j2];
            this.data[i2Plus1] = this.data[i2Plus1] + this.data[j2Plus1];
            this.data[j2Plus1] = this.data[i2Plus1] - this.data[j2Plus1];
            this.data[i2Plus1] = this.data[i2Plus1] - this.data[j2Plus1];
        }
    }

    public exportLegacy(): IPoint[] {
        const result: IPoint[] = [];
        const pointCount: number = this.pointCount;
        let i: number = 0;

        for (i = 0; i < pointCount; ++i) {
            result.push(this.at(i).export());
        }

        return result;
    }

    private getPointOffset(index: number): number {
        return this.offset + (index << 1);
    }

    private calculateBounds(): void {
        if (this.isBroken) {
            return;
        }

        const point1: Point = Point.from(this.first);
        const point2: Point = Point.from(this.last);

        this.closed = point1.almostEqual(point2);

        const pointCount: number = this.pointCount;
        let i: number = 0;
        let point: Point = null;

        for (i = 0; i < pointCount; ++i) {
            point = this.at(i);
            point1.min(point);
            point2.max(point);
        }

        this.rectangle = true;

        for (i = 0; i < this.pointCount; ++i) {
            point = this.at(i);

            if (
                !(
                    (almostEqual(point.x, point1.x) || almostEqual(point.x, point2.x)) &&
                    (almostEqual(point.y, point1.y) || almostEqual(point.y, point2.y))
                )
            ) {
                this.rectangle = false;
                break;
            }
        }

        point2.sub(point1);

        this.tmpPoint.bind(this.rectData, 0).update(point1);
        this.tmpPoint.bind(this.rectData, 2).update(point2);
    }

    public get length(): number {
        const offset: number = this.closedDirty ? 1 : 0;

        return this.pointCount + offset;
    }

    public get first(): Point {
        return this.at(0);
    }

    public get last(): Point {
        return this.at(cycleIndex(this.length, this.pointCount, -1));
    }

    public get children(): Polygon[] {
        return this.innerChildren;
    }

    public get childrCount(): number {
        return this.innerChildren.length;
    }

    public get hasChildren(): boolean {
        return this.childrCount !== 0;
    }

    public get isBroken(): boolean {
        return this.length < 3;
    }

    public get isClosed(): boolean {
        return this.closed || this.closedDirty;
    }

    public get isRectangle(): boolean {
        return this.rectangle;
    }

    // returns the area of the polygon, assuming no self-intersections
    // a negative area indicates counter-clockwise winding direction
    public get area(): number {
        const pointCount = this.pointCount;
        const prevPoint: Point = Point.zero();
        const currPoint: Point = Point.zero();
        let result: number = 0;
        let i: number = 0;

        for (i = 0; i < pointCount; ++i) {
            prevPoint.update(this.at(cycleIndex(i, pointCount, -1)));
            currPoint.update(this.at(i));
            result += (prevPoint.x + currPoint.x) * (prevPoint.y - currPoint.y);
        }

        return 0.5 * result;
    }

    public get position(): Point {
        return this.tmpPoint.bind(this.rectData, 0);
    }

    public get size(): Point {
        return this.tmpPoint.bind(this.rectData, 2);
    }

    public static fromLegacy(data: IPolygon | IPoint[]): Polygon {
        const result = new Polygon();

        result.reset(data);

        return result;
    }
}

File: packages/wasm-nesting/src/geometry/polygon.rs
use crate::geometry::bound_rect::BoundRect;
use crate::geometry::point::Point;
use crate::utils::math::cycle_index;
use crate::utils::number::Number;

pub struct Polygon<T: Number> {
    buffer: Option<Box<[T]>>,
    offset: usize,
    point_count: usize,
    closed: bool,
    closed_dirty: bool,
    calculate_bounds_dirty: bool,
    rectangle: bool,
    bounds: BoundRect<T>,
    cursor: Point<T>,
}

impl<T: Number> Polygon<T> {
    pub fn new() -> Self {
        Self {
            buffer: None,
            offset: 0,
            point_count: 0,
            closed: false,
            closed_dirty: false,
            rectangle: false,
            calculate_bounds_dirty: true,
            bounds: BoundRect::new(T::zero(), T::zero(), T::zero(), T::zero()),
            cursor: Point::new(None, None),
        }
    }

    pub unsafe fn bind(&mut self, buffer: Box<[T]>, offset: usize, point_count: usize) {
        self.buffer = Some(buffer);
        self.offset = offset;
        self.point_count = point_count;
        self.closed_dirty = false;
        self.closed = false;
        self.rectangle = false;
        self.calculate_bounds_dirty = true;
    }

    pub fn clean(&mut self) {
        self.buffer = None;
        self.point_count = 0;
        self.offset = 0;
        self.closed_dirty = false;
        self.closed = false;
        self.rectangle = false;
        self.calculate_bounds_dirty = true;
    }

    pub unsafe fn at(&mut self, index: usize) -> *const Point<T> {
        if let Some(ref buf) = self.buffer {
            let point_index = cycle_index(index, self.point_count, 0);
            let idx = self.offset + (point_index << 1);

            self.cursor.set(buf[idx], buf[idx + 1])
        } else {
            self.cursor.set(T::zero(), T::zero())
        }
    }

    pub unsafe fn first(&mut self) -> *const Point<T> {
        self.at(0)
    }

    pub unsafe fn last(&mut self) -> *const Point<T> {
        let idx = cycle_index(self.length(), self.point_count, -1);
        self.at(idx)
    }

    pub fn length(&self) -> usize {
        let offset = if self.closed_dirty { 1 } else { 0 };
        self.point_count + offset
    }

    pub fn is_broken(&self) -> bool {
        self.length() < 3
    }

    pub fn is_closed(&self) -> bool {
        self.closed || self.closed_dirty
    }

    pub unsafe fn is_rectangle(&mut self) -> bool {
        self.calculate_bounds();
        self.rectangle
    }

    pub unsafe fn area(&mut self) -> f64 {
        return Number::polygon_area(self.buffer.as_ref().unwrap());
    }

    pub unsafe fn abs_area(&mut self) -> f64 {
        self.area().abs()
    }

    pub unsafe fn position(&mut self) -> *const Point<T> {
        self.calculate_bounds();
        self.bounds.position()
    }

    pub unsafe fn size(&mut self) -> *const Point<T> {
        self.calculate_bounds();
        self.bounds.size()
    }

    pub unsafe fn export_bounds(&mut self) -> BoundRect<T> {
        self.calculate_bounds();
        self.bounds.clone()
    }

    unsafe fn calculate_bounds(&mut self) {
        if self.is_broken() || !self.calculate_bounds_dirty {
            return;
        }

        self.calculate_bounds_dirty = false;

        if let Some(ref buf) = self.buffer {
            // Use Number::calculate_bounds to get [x, y, width, height]
            let bounds = T::calculate_bounds(buf, self.offset, self.point_count);

            let min_point = Point::new(Some(bounds[0]), Some(bounds[1]));
            let size_point = Point::new(Some(bounds[2]), Some(bounds[3]));

            // Check if polygon is closed
            let first = Point::from(self.first());
            let last = Point::from(self.last());
            self.closed = first.almost_equal(&last as *const Point<T>, None);

            // Check if polygon is a rectangle
            self.rectangle = true;
            let max_x = bounds[0] + bounds[2];
            let max_y = bounds[1] + bounds[3];

            for i in 0..self.point_count {
                let pt = self.at(i);
                let px = (*pt).x;
                let py = (*pt).y;

                if !((px.almost_equal(bounds[0], None) || px.almost_equal(max_x, None))
                    && (py.almost_equal(bounds[1], None) || py.almost_equal(max_y, None)))
                {
                    self.rectangle = false;
                    break;
                }
            }

            self.bounds.update(&min_point, &size_point);
        }
    }

    pub fn close(&mut self) {
        if self.is_closed() {
            return;
        }

        self.closed_dirty = true;
    }

    pub fn reverse(&mut self) {
        if self.buffer.is_none() {
            return;
        }

        Number::reverse_polygon(self.buffer.as_mut().unwrap(), self.offset, self.point_count);
    }

    pub unsafe fn normalize(&mut self) -> Option<Box<[T]>> {
        if self.buffer.is_none() {
            return None;
        }

        let first = Point::from(self.first());
        let mut last = Point::from(self.last());
        let mut point_count = self.point_count;

        while first.almost_equal(&last, None) && point_count > 1 {
            point_count -= 1;
            last.update(self.at(point_count - 1));
        }

        let buf = self.buffer.as_mut().unwrap();

        if self.point_count != point_count {
            self.point_count = point_count;
            let start = self.offset;
            let end = self.offset + (point_count << 1);
            let slice = &buf[start..end];
            let boxed = slice.to_vec().into_boxed_slice();
            self.buffer = Some(boxed);
        }

        if self.area() > 0.0 {
            self.reverse();
        }

        self.buffer.as_ref().map(|b| b.clone())
    }

    pub unsafe fn point_in(
        &mut self,
        point: *const Point<T>,
        offset: Option<*const Point<T>>,
    ) -> Option<bool> {
        if self.is_broken() {
            return None;
        }

        let inner_point = Point::<T>::from(point);
        let mut curr_point = Point::<T>::new(None, None);
        let mut prev_point = Point::<T>::new(None, None);

        let point_count = self.point_count as usize;

        let mut inside = false;
        for i in 0..point_count {
            curr_point.update(self.at(i));

            let prev_idx = cycle_index(i, point_count, -1);

            prev_point.update(self.at(prev_idx));

            if let Some(off) = offset {
                curr_point.add(off);
                prev_point.add(off);
            }

            // Якщо точка точно лежить на вершині або на сегменті — повертаємо None
            if curr_point.almost_equal(&inner_point, None)
                || inner_point.on_segment(&curr_point, &prev_point)
            {
                return None;
            }

            if curr_point.almost_equal(&prev_point, None) {
                continue;
            }

            let curr_y_gt = curr_point.y > inner_point.y;
            let prev_y_gt = prev_point.y > inner_point.y;

            if curr_y_gt != prev_y_gt {
                let inter_x = inner_point.interpolate_x(&prev_point, &curr_point);
                if inner_point.x < inter_x {
                    inside = !inside;
                }
            }
        }

        Some(inside)
    }
}
